### ISSUE TITLE
Consolidate files listing dependencies

### ISSUE THREAD

**Issue by adamjstewart** [microsoft/torchgeo#140]
Currently, the following files all list torchgeo's dependencies:

* `setup.cfg`
* `requirements.txt`
* `docs/requirements.txt`
* `spack.yaml`
* `environment.yml`
* `.github/workflows/{docs/release/style/tests}.yaml`
* `.pre-commit-config.yaml`

Whenever someone needs to add a new dependency to torchgeo, they need to add it to most of these files, making it easy to forget and difficult to keep track of. We should try to consolidate these. Here are some suggestions:

### `setup.cfg`

We may be able to use `requirements.txt` instead of `setup.cfg` by using something like pbr or setuptools-scm. But see below for another option.

### `requirements.txt`

I don't think this file is necessary. `pip install .[all]` in the root directory should read the dependencies from `setup.cfg` and provide much of the same support as `pip install -r requirements.txt`. I think `pip install git+https://github.com/microsoft/torchgeo` may also work.

### `docs/requirements.txt`

This can be removed pending https://github.com/pytorch/pytorch_sphinx_theme/issues/143

### `spack.yaml`

This can be removed by adding a `py-torchgeo` package to Spack. It can install both stable releases (once we release) as well as development versions.

### `environment.yml`

I'm not sure how to remove this one. Can conda-forge be used to install development versions?

### `.github/workflows/{docs/release/style/tests}.yaml`

If we `pip install` all of our dependencies from `requirements.txt`, it will be slower than only installing the deps for a specific tests, but we won't have to duplicate info as much.

### `.pre-commit-config.yaml`

No ideas, the developers don't seem interested in sharing configuration files with any other tool: https://github.com/pre-commit/pre-commit/issues/1165

**Comment by ashnair1** [comment]
[Poetry](https://github.com/python-poetry/poetry) might be a good solution for this. 

Advantages:
- Integrates well with `pyproject.toml`. In fact poetry considers `pyproject.toml` the single source of truth.
- Once integrated we won't need `requirements.txt`, `environment.yaml` and `setup.cfg`. 
- Also makes it easy to publish to pypi.

Disadvantages (maybe?):
- Dependencies will need to be moved to a single location - `pyproject.toml`
- Switching to poetry might be non-trivial for a large project like this.

Personally, I've switched from using conda to poetry and have found it much better. If there is interest I can draft a PR.

**Comment by adamjstewart** [comment]
Yes, I've been keeping an eye on alternative build tools like poetry and flit. So far I haven't seen a significant advantage compared to setuptools and `setup.cfg`.

> Supports [optional dependencies](https://python-poetry.org/docs/pyproject/#extras) and recently added support for python-poetry/poetry/pull/4260. Could be useful for installing specific group of deps for specific tests.

Setuptools has always had support for this, so I don't see this as an advantage.

> Personally, I've switched from using conda to poetry in my projects and have found it much better but YMMV.

How does `poetry` replace `conda`?

**Comment by ashnair1** [comment]
Perhaps better is too strong a word. Convenient would be more appropriate. Poetry doesn't replace conda. conda manages environments and resolves dependencies whereas poetry only does the latter. The key here is how a project uses conda. Is it used for environment management? Installing specific non-python libraries? Both?

I mainly used conda for environment management and I preferred poetry because:


- Manual update and maintenance of environment.yml
With conda I needed to maintain an environment.yml file and had to manually update it for every package I added. 
With poetry I can keep adding dependencies via `poetry add` which updates `pyproject.toml` and the `poetry.lock` file (for transitive dependencies) and go about my work. Also one less file for me to maintain.


- conda is generally slower.
This isn't a knock against conda since it's scope extends beyond python and python packages. But due to this, it takes a while for dependencies to be resolved. poetry is a dependency manager for python and I've found it to be faster to set up an environment when I work with python projects. 

Misc issues:
- Need to keep track of conda-forge and pypi versions. 
- Cross platform environments


Poetry  and pyenv together gives me enough [environment management](https://python-poetry.org/docs/managing-environments/) without the overhead of maintaining an additional file of dependencies and tracking a separate package registry (`conda-forge`). Here I prefer to use two specialised tools over a single massive one.


That being said, for projects that utilise non-python libraries (like `GDAL` for example), I would still use conda.
